Fix part of #2863: Create ParamSpecs and ParamSpec objects.

brianrodri opened the pull request:
Fix part of #2863: Create ParamSpecs and ParamSpec objects.

brianrodri committed:
Merged to upstream/develop

brianrodri committed:
Remove mysterious help directory

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Introduces ParamSpecsObjectFactory for working with ParamSpecs through
an object interface.

brianrodri committed:
Introduces ParamSpecs to define a collection of parameter
specifications.

brianrodri committed:
Introduce an individual ParamSpec for parameters.

brianrodri committed:
Use helper createDefault function to create default parameter specs.

brianrodri committed:
Introduce differentiation between backend dict and the real ParamSpecs/ParamSpec objects.

brianrodri committed:
Make ParamSpecs python-representation more extensible for future object types.

brianrodri committed:
Use single-quotes as required by style guide.

brianrodri committed:
Fix identifier name to camel case.

brianrodri committed:
Fix whitespace lint errors.

brianrodri committed:
Fix file to end with single newline

brianrodri committed:
Fixes files to end with exactly one newline.

brianrodri committed:
Use helper method paramNames() instead.

brianrodri committed:
Improve formatting by using temporary variables.

brianrodri committed:
Fix indentation errors.

brianrodri committed:
Add convenience function "addParamIfNew" to simplify many usages.

brianrodri committed:
Use this where valid for readability.

brianrodri committed:
Update validation test to use new match against new error string.

brianrodri committed:
Improve formatting of validation error

brianrodri committed:
Fix indentation.

brianrodri committed:
Remove index argument since it isn't used.

brianrodri committed:
Fix typo in which param name is used.

tjiang11 commented on core/templates/dev/head/domain/exploration/ParamSpecObjectFactory.js:
```diff
@@ -0,0 +1,40 @@
+// Copyright 2015 The Oppia Authors. All Rights Reserved.
```
2017 :)

brianrodri replied:
Done.

tjiang11 commented on core/templates/dev/head/domain/exploration/ParamSpecsObjectFactory.js:
```diff
@@ -0,0 +1,62 @@
+// Copyright 2015 The Oppia Authors. All Rights Reserved.
```
Ditto

brianrodri replied:
Done.

tjiang11 commented on core/templates/dev/head/domain/exploration/ParamSpecsObjectFactory.js:
```diff
@@ -0,0 +1,62 @@
+    ParamSpecs.prototype.toBackendDict = function() {
+      paramSpecsBackendDict = {};
+      that = this;
+      this.paramNames().map(function(paramName) {
```
forEach may be better since you're not using the returned list.

brianrodri replied:
Done.

tjiang11 commented on core/templates/dev/head/domain/exploration/ParamSpecsObjectFactory.js:
```diff
@@ -0,0 +1,62 @@
+      this.paramSpecs = paramSpecs;
+    };
+
+    ParamSpecs.prototype.paramNames = function() {
```
paramNames --> getParamNames

brianrodri replied:
Done.

tjiang11 commented on core/templates/dev/head/domain/exploration/ParamSpecsObjectFactory.js:
```diff
@@ -0,0 +1,62 @@
+      return Object.keys(this.paramSpecs);
+    };
+
+    ParamSpecs.prototype.addParamIfNew = function(paramName, paramSpec) {
```
It'd be nice if you could include a brief karma test for this.

brianrodri replied:
Done.

tjiang11 commented on core/templates/dev/head/domain/exploration/ParamSpecsObjectFactory.js:
```diff
@@ -0,0 +1,62 @@
+    };
+
+    ParamSpecs.createFromBackendDict = function(paramSpecsBackendDict) {
+      paramSpecs = {};
```
`var paramSpecs = {};` (make local)

brianrodri replied:
Done.

tjiang11 commented on core/templates/dev/head/domain/exploration/ParamSpecsObjectFactory.js:
```diff
@@ -0,0 +1,62 @@
+    };
+
+    ParamSpecs.createFromBackendDict = function(paramSpecsBackendDict) {
+      paramSpecs = {};
```
Can you call this paramSpecObjectsDict or similar? Just want to somehow distinguish it from the parent class.

brianrodri replied:
Done.

tjiang11 replied:
@brianrodri Oh, I was actually referring to just this local variable, not the object property. But it's fine either way, really. I'll leave it up to you.

brianrodri replied:
Done, decided on paramDict and it looks a lot better!

tjiang11 reviewed (changes requested):
Thanks @brianrodri !

brianrodri committed:
Fix copyright dates.

brianrodri committed:
Rename paramNames to getParamNames.

brianrodri committed:
Use forEach instead of map.

brianrodri committed:
Make paramSpecs a local object.

brianrodri committed:
Rename internal variable to differentiate from class name.

brianrodri committed:
Rename internal object to differentiate from class name.

brianrodri committed:
Rename internal object to differentiate from class name.

brianrodri committed:
Adds unit tests for ParamSpecsObjectFactory.

brianrodri committed:
Adds accessor for underlying object dict of ParamSpecs.

brianrodri committed:
Uses new accessor methods for param specs.

brianrodri committed:
Remove unused ParamSpecObjectFactory injection.

tjiang11 commented:
Lgtm! Just one comment (see last outdated comment)

brianrodri committed:
Rename paramSpecsObjectDict to plain old paramSpecsDict for easier readability.

brianrodri committed:
Use paramDict instead, still retains useful info.

brianrodri committed:
Make regex matches less rigid.

brianrodri committed:
Make variables local.

brianrodri committed:
Fixes typo in parameter name.

brianrodri committed:
Fix parameter name.

brianrodri committed:
Add documentation to new classes.

brianrodri committed:
Fix indentation errors.

brianrodri committed:
Documentation for required arguments.

brianrodri committed:
Fix documentation and denote required arguments.

brianrodri committed:
Add accessor for paramDict

brianrodri committed:
Use forEach instead of for in, fixing bug which looped through array indicies instead of values.

brianrodri committed:
Fix settings_tab to use correct member.

brianrodri committed:
Use accessor method for objType

brianrodri committed:
Make note of good point to introduce Object Types

seanlip commented:
Hm, there seem to be a lot of commits since @tjiang11's last approval of this PR. Should a follow-up review be done?

brianrodri commented:
Yes please! There were a few bugs I fixed and I've also added some documentation/tests, so there's plenty of content worth reviewing again.

@tjiang11, could you PTAL?

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into param-specs

brianrodri committed:
Remove unused injection stuff from tests.

brianrodri committed:
Unnest specs

brianrodri committed:
Use toBe(true/false) instead of toBeTruthy/Falsy

brianrodri commented:
I've decided that ParamTypes should be introduced at the same time, so I'll close this PR and defer to PR#3870 instead!
